### ISSUE TITLE
Fallback to IPv4 when IPv6 connexion fails is broken

### ISSUE THREAD

**Issue by smortex** [nats-io/pure-ruby-nats#21]
gem version: nats-pure-0.2.4

#### Context

A dual-stacked machine I manage lost it's IPv6 connectivity and the nats client running on it was unable to recover by connecting to the dual-stacked natsd server over IPv4.

The client still has an IPv6 address, IPv6 routes and all (the network that connects it to the internet over IPv6 is broken, but IPv6 itself is working correctly for local stuff), so `getaddrinfo` returns both an IPv6 and an IPv4 address for the remote server.

#### Problem description

When nats-pure detects that the connection over IPv6 fails (first address), it does not try to connect using IPv4 (second address) address, but reties to connect using IPv6 over and over.

#### Expected behavior

When a connection fails or gets broken, nats-pure should attempt to connect using all remaining IP addresses before looping and retrying to connect using the first one.

#### Work-around

While the public network is broken, I could successfully connect by changing `AF_UNSPEC` to `PF_INET` in 
https://github.com/nats-io/pure-ruby-nats/blob/master/lib/nats/io/client.rb#L1044.

**Comment by wallyqs** [comment]
Thanks for raising the issue, though I haven't been able to reproduce locally so far and the client is trying both ipv6 and ipv4 addresses: https://github.com/nats-io/pure-ruby-nats/blob/4a60b9fe1fbfcd7bc4b62feb4d9aca7e728b9e00/lib/nats/io/client.rb#L1043-L1057

I wonder if it was an issue around maximum number of reconnect attempts? If the client cannot connect to the server more than `:max_reconnect_attempts` times it will discard the server when attempting to reconnect, to prevent this it is possible to set `:max_reconnect_attempts => -1` to reconnect infinitely though.  If running into this you could confirm by defining the `on_close` callback which is triggered once the client runs out of servers.
```
nats.on_close do |e|
  puts "Closing... #{nats.last_error}"
end
```

**Comment by smortex** [comment]
The networking problem has been fixed, but I only realized later that this client was not the only one to be isolated from an IPv6 point of view, however, it was the _only one_ affected by the fallback problem :thinking:.

All nodes run the same version of nats-pure.

I guess simulating the problem would be as easy as blocking all IPv6 using a firewall on a node between the machines, but I am not sure to be able to allocate time for this, so closing for now.  If I have further information, I'll follow-up to this PR, and if the source problem is found, re-open it.

Thanks @wallyqs !
